File: pf-core/src/main/java/com/purdynet/prices/PriceRecord.java
package com.purdynet.prices;

import java.math.BigDecimal;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;

/**
 * Created with IntelliJ IDEA.
 * User: dnpurdy
 * Date: 8/28/13
 * Time: 7:13 PM
 * To change this template use File | Settings | File Templates.
 */
public class PriceRecord implements Comparable<PriceRecord>
{
    private Date date;
    private Integer year;
    private Integer month;
    private Integer day;
    private BigDecimal price;

    public PriceRecord(String dateString, BigDecimal price)
    {
        final SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd");
        final Calendar c = Calendar.getInstance();
        try {
            date = sdf.parse(dateString);
            c.setTime(date);
            this.year = c.get(Calendar.YEAR);
            this.month = c.get(Calendar.MONTH)+1;
            this.day = c.get(Calendar.DAY_OF_MONTH);
            this.price = price;
        }
        catch (ParseException e) {
            e.printStackTrace();
        }
    }

    public Integer getYear() {
        return year;
    }

    public Integer getMonth() {
        return month;
    }

    public Integer getDay() {
        return day;
    }

    public BigDecimal getPrice() {
        return price;
    }

    @Override
    public String toString() {
        return getDateCode()+": "+price.toString();
    }

    @Override
    public int compareTo(PriceRecord o) {
        if(this.year.compareTo(o.getYear())!=0) return this.year.compareTo(o.getYear());
        else if(this.month.compareTo(o.getMonth())!=0) return this.month.compareTo(o.getMonth());
        else return this.day.compareTo(o.getDay());
    }

    public String getDateCode()
    {
        SimpleDateFormat dt1 = new SimpleDateFormat("yyyy-MM-dd");
        return dt1.format(date);
    }
}

File: pf-core/src/main/scala/com/purdynet/prices/PriceRecord.scala
package com.purdynet.prices

import java.text.SimpleDateFormat
import java.util.Date
import java.util.concurrent.TimeUnit

case class PriceRecord(date: Date, price: BigDecimal) extends Ordered[PriceRecord] {
  def this(dateString: String, price: BigDecimal) {
    this(PriceRecord.parseDateString(dateString), price)
  }

  @deprecated
  def getDateStr: String = PriceRecord.formatDateString(date)

  @deprecated
  def getJavaPrice: java.math.BigDecimal = new java.math.BigDecimal(price.longValue())

  def daysBetween(o: PriceRecord): Long = {
    val diff = o.date.getTime - this.date.getTime
    TimeUnit.DAYS.convert(diff, TimeUnit.MILLISECONDS)
  }

  override def toString: String = getDateStr + ": " + price.toString

  override def compare(that: PriceRecord): Int = this.date.compareTo(that.date)
}

object PriceRecord {
  final val sdf = new SimpleDateFormat("yyyy-MM-dd")

  def parseDateString(s: String): Date = sdf.parse(s)

  def formatDateString(d: Date): String = sdf.format(d)
}